File: src/main/java/wraith/fabricaeexnihilo/util/EnchantmentTagManager.java
package wraith.fabricaeexnihilo.util;

import net.devtech.arrp.api.RuntimeResourcePack;
import net.devtech.arrp.json.tags.JTag;
import net.fabricmc.fabric.api.tag.TagFactory;
import net.minecraft.enchantment.Enchantment;
import net.minecraft.enchantment.EnchantmentLevelEntry;
import net.minecraft.enchantment.Enchantments;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.tag.Tag;
import net.minecraft.util.Identifier;
import net.minecraft.util.registry.Registry;
import wraith.fabricaeexnihilo.FabricaeExNihilo;
import wraith.fabricaeexnihilo.modules.ModBlocks;
import wraith.fabricaeexnihilo.modules.ModItems;
import wraith.fabricaeexnihilo.modules.ModTags;

import java.util.HashMap;
import java.util.List;

import static wraith.fabricaeexnihilo.FabricaeExNihilo.id;

public final class EnchantmentTagManager {
    
    private EnchantmentTagManager() {
    }
    
    public static List<EnchantmentLevelEntry> getHighestApplicableEnchantmentsAtPower(int power, ItemStack stack, boolean hasTreasure) {
        return getApplicableEnchantments(stack, hasTreasure).stream().map(enchantmentLevelEntry -> getHighestLevelAtPower(enchantmentLevelEntry, power)).toList();
    }
    
    
    public static Identifier getTagIDForEnchantment(Enchantment enchantment) {
        var enchID = RegistryUtils.getId(enchantment);
        if (enchID == null) {
            enchID = new Identifier("empty");
        }
        return id("enchantments/" + enchID.getNamespace() + "/" + enchID.getPath());
    }
    
    public static boolean itemIsAcceptableForEnchantment(ItemStack stack, Enchantment enchantment) {
        return TagFactory.ITEM.create(getTagIDForEnchantment(enchantment)).contains(stack.getItem());
    }
    
    private static Tag.Identified<Item> getTagForEnchantment(Enchantment enchantment) {
        return TagFactory.ITEM.create(getTagIDForEnchantment(enchantment));
    }
    
    /**
     * @return A list of enchantments for which this item is tagged with the item tag
     * fabricaeexnihilo:enchantments/modid/enchantment_id
     */
    public static List<Enchantment> getApplicableEnchantments(ItemStack stack, boolean hasTreasure) {
        return Registry.ENCHANTMENT.stream()
                .filter(enchantment -> hasTreasure || !enchantment.isTreasure())
                .filter(enchantment -> getTagForEnchantment(enchantment).contains(stack.getItem()))
                .toList();
    }
    
    public static void generateDefaultTags(RuntimeResourcePack resourcePack) {
        
        JTag tag = JTag.tag();
        if (FabricaeExNihilo.CONFIG.modules.crucibles.efficiency) {
            ModTags.addAllTags(tag, ModBlocks.CRUCIBLES.keySet());
        }
        if (FabricaeExNihilo.CONFIG.modules.barrels.efficiency) {
            ModTags.addAllTags(tag, ModBlocks.BARRELS.keySet());
        }
        if (FabricaeExNihilo.CONFIG.modules.sieves.efficiency) {
            ModTags.addAllTags(tag, ModItems.MESHES.keySet());
        }
        resourcePack.addTag(id("items/" + getTagIDForEnchantment(Enchantments.EFFICIENCY).getPath()), tag);
        
        tag = JTag.tag();
        if (FabricaeExNihilo.CONFIG.modules.barrels.thorns) {
            ModTags.addAllTags(tag, ModBlocks.BARRELS.keySet());
        }
        resourcePack.addTag(id("items/" + getTagIDForEnchantment(Enchantments.THORNS).getPath()), tag);
        
        tag = JTag.tag();
        if (FabricaeExNihilo.CONFIG.modules.crucibles.fireAspect) {
            ModTags.addAllTags(tag, ModBlocks.CRUCIBLES.keySet());
        }
        resourcePack.addTag(id("items/" + getTagIDForEnchantment(Enchantments.FIRE_ASPECT).getPath()), tag);
        
        tag = JTag.tag();
        if (FabricaeExNihilo.CONFIG.modules.sieves.fortune) {
            ModTags.addAllTags(tag, ModItems.MESHES.keySet());
        }
        resourcePack.addTag(id("items/" + getTagIDForEnchantment(Enchantments.FORTUNE).getPath()), tag);
    }
    
    public static List<EnchantmentLevelEntry> mergeInfoLists(List<EnchantmentLevelEntry> first, List<EnchantmentLevelEntry> second) {
        var mapping = new HashMap<Enchantment, Integer>();
        first.forEach(enchantment -> mapping.put(enchantment.enchantment, Math.max(mapping.getOrDefault(enchantment.enchantment, 0), enchantment.level)));
        second.forEach(enchantment -> mapping.put(enchantment.enchantment, Math.max(mapping.getOrDefault(enchantment.enchantment, 0), enchantment.level)));
        return mapping.entrySet().stream().map(entry -> new EnchantmentLevelEntry(entry.getKey(), entry.getValue())).toList();
    }
    
    public static EnchantmentLevelEntry getHighestLevelAtPower(Enchantment enchantment, int power) {
        for (var level = enchantment.getMaxLevel(); level >= enchantment.getMinLevel() - 1; level--) {
            if (power >= enchantment.getMinPower(level) && power <= enchantment.getMaxPower(level)) {
                return new EnchantmentLevelEntry(enchantment, level);
            }
        }
        return new EnchantmentLevelEntry(enchantment, 1);
    }
    
}

File: src/main/kotlin/exnihilofabrico/impl/EnchantmentTagManager.kt
package exnihilofabrico.impl

import com.swordglowsblue.artifice.api.ArtificeResourcePack
import exnihilofabrico.ExNihiloFabrico
import exnihilofabrico.api.registry.ExNihiloRegistries
import exnihilofabrico.id
import exnihilofabrico.modules.ModBlocks
import net.fabricmc.fabric.api.tag.TagRegistry
import net.minecraft.enchantment.Enchantment
import net.minecraft.enchantment.Enchantments
import net.minecraft.enchantment.InfoEnchantment
import net.minecraft.item.ItemStack
import net.minecraft.util.Identifier
import net.minecraft.util.registry.Registry

object EnchantmentTagManager {
    fun getHighestApplicableEnchantmentsAtPower(power: Int, stack: ItemStack, hasTreasure: Boolean): List<InfoEnchantment> {
        val enchantments = getApplicableEnchantments(stack, hasTreasure)

        return enchantments.map { it.getHighestLevelAtPower(power) }
    }


    fun getTagIDForEnchantment(enchantment: Enchantment): Identifier {
        val enchID = Registry.ENCHANTMENT.getId(enchantment) ?: Identifier("empty")
        return id("enchantments/${enchID.namespace}/${enchID.path}")
    }

    fun itemIsAcceptableForEnchantment(stack: ItemStack, enchantment: Enchantment): Boolean {
        return stack.item.isIn(TagRegistry.item(
            getTagIDForEnchantment(
                enchantment
            )
        ))
    }

    /**
     * @return  A list of enchantments for which this item is tagged with the item tag
     *          exnihilofabrico:enchantments/modid/enchantment_id
     */
    fun getApplicableEnchantments(stack: ItemStack, hasTreasure: Boolean): List<Enchantment> =
        Registry.ENCHANTMENT
            .filter { hasTreasure || !it.isTreasure }
            .filter { stack.item.isIn(TagRegistry.item(
                getTagIDForEnchantment(
                    it
                )
            )) }


    fun generateDefaultTags(builder: ArtificeResourcePack.ServerResourcePackBuilder) {
        builder.addItemTag(getTagIDForEnchantment(Enchantments.EFFICIENCY)) { tag ->
            if(ExNihiloFabrico.config.modules.crucibles.efficiency)
                tag.values(*ModBlocks.CRUCIBLES.keys.toTypedArray())
            if(ExNihiloFabrico.config.modules.barrels.efficiency)
                tag.values(*ModBlocks.BARRELS.keys.toTypedArray())
            if(ExNihiloFabrico.config.modules.sieves.efficiency)
                tag.values(*ExNihiloRegistries.MESH.getAll().map { it.identifier }.toTypedArray())
        }
        builder.addItemTag(getTagIDForEnchantment(Enchantments.FIRE_ASPECT)) { tag ->
            if(ExNihiloFabrico.config.modules.crucibles.fireAspect)
                tag.values(*ModBlocks.CRUCIBLES.keys.toTypedArray())
        }
        builder.addItemTag(getTagIDForEnchantment(Enchantments.FORTUNE)) { tag ->
            if(ExNihiloFabrico.config.modules.sieves.fortune)
                tag.values(*ExNihiloRegistries.MESH.getAll().map { it.identifier }.toTypedArray())
        }
    }

    fun mergeInfoLists(first: List<InfoEnchantment>, second: List<InfoEnchantment>): List<InfoEnchantment> {
        val mapping = mutableMapOf<Enchantment, Int>()
        for(set in listOf(first, second))
            set.forEach {
                mapping[it.enchantment] = maxOf(mapping.getOrDefault(it.enchantment, 0), it.level)
            }
        return mapping.map { (k,v) -> InfoEnchantment(k,v) }
    }
}

fun Enchantment.getHighestLevelAtPower(power: Int): InfoEnchantment {
    return (this.maximumLevel downTo this.minimumLevel-1)
        .filter { power >= this.getMinimumPower(it) && power <= this.getMaximumPower(it) }
        .map { InfoEnchantment(this, it) }.first()
}